src/fluree/db/ledger/indexing.clj

```diff
@@ -15,390 +17,374 @@
-                                      :size      novelty-size
-                                      :completed 0})
-           ^Instant start-time (Instant/now)]
-       (log/info (str "Index Update begin at: " start-time) {:network      network
-                                                             :dbid         dbid
-                                                             :t            t
-                                                             :block        block
-                                                             :novelty-size novelty-size})
-       (if-not db-dirty?
-         db
-         (let [spot-ch    (index-root db progress :spot)    ;; indexes run in parallel
-               psot-ch    (index-root db progress :psot)
-               post-ch    (index-root db progress :post remove-preds)
-               opst-ch    (index-root db progress :opst)
-               indexed-db (-> db
-                              (assoc :spot (<? spot-ch)
-                                     :psot (<? psot-ch)
-                                     :post (<? post-ch)
-                                     :opst (<? opst-ch))
-                              (update-in [:novelty :spot] empty) ;; retain sort order of indexes
-                              (update-in [:novelty :psot] empty)
-                              (update-in [:novelty :post] empty)
-                              (update-in [:novelty :opst] empty)
-                              (assoc-in [:novelty :size] 0)
-                              (assoc-in [:stats :indexed] block))]
-           ;; wait until confirmed writes before returning
-           ;; TODO - ideally issue garbage/root writes to RAFT together as a tx, currently requires waiting for both through raft sync
-           (<? (storage/write-garbage indexed-db @progress))
-           (<? (storage/write-db-root indexed-db ecount))
-           (log/info (str "Index Update end at: " (Instant/now))
-                     {:network      network
-                      :dbid         dbid
-                      :block        block
-                      :t            t
-                      :idx-duration (- (.toEpochMilli ^Instant (Instant/now))
-                                       (.toEpochMilli start-time))})
-           indexed-db))))))
-
+    (let [start-time   (Instant/now)
+          novelty-size (:size novelty)
+          init-stats   {:network      network
+                        :dbid         dbid
+                        :t            t
+                        :block        block
+                        :novelty-size novelty-size
+                        :start-time   start-time}]
+      (if (or (dirty? db)
+              (seq remove-preds))
+        (do (log/info "Refreshing Index:" init-stats)
+            (let [error-ch   (chan)
+                  refresh-ch (refresh-all db remove-preds error-ch)]
+              (async/alt!
```

Comment:
I like this pattern of a separate error chan and a results chan - it separates concerns nicely!